app/src/test/java/com/github/raulccabreu/redukt/BaseAnnotatedMiddlewareTest.kt

```diff
@@ -193,6 +179,246 @@ class BaseAnnotatedMiddlewareTest {
         }
     }
 
+    @Test
+    fun whenUseBeforeActions() {
+        var result: Pair<String, String>? = null
+        val signal = CountDownLatch(1)
+
+        val beforeMiddleware = ValidMiddlewareActions { state, actionName ->
+            result = Pair(state, actionName)
+            signal.countDown()
+        }
+
+        val redukt = Redukt("initial")
+        redukt.middlewares.add(beforeMiddleware)
+
+        redukt.dispatch(Action("valid", "new state"), false)
+
+        signal.await()
+
+        Assert.assertEquals("initial", result?.first)
+        Assert.assertEquals("valid", result?.second)
+
+        redukt.stop()
+    }
+
+    @Test
+    fun whenUseAfterActions() {
+        var result: Pair<String, String>? = null
+        val signal = CountDownLatch(1)
+
+        val beforeMiddleware = ValidMiddlewareActions ( afterCallback = { state, actionName ->
+            result = Pair(state, actionName)
+            signal.countDown()
+        })
+
+        val redukt = Redukt("initial")
+        redukt.middlewares.add(beforeMiddleware)
+
+        redukt.dispatch(Action("valid", "new state"), false)
+
+        signal.await()
+
+        Assert.assertEquals("initial", result?.first)
+        Assert.assertEquals("valid", result?.second)
+
+        redukt.stop()
+    }
+
+    @Test
+    fun whenUseBeforeAfterActions() {
+        var resultBefore: Pair<String, String>? = null
+        var resultAfter: Pair<String, String>? = null
+        val signal = CountDownLatch(2)
+
+        val beforeMiddleware = ValidMiddlewareActions ({ state, actionName ->
```

Comment:
Suggestion: Maybe a class att (the callbacks implementation are similar)? With more generic name? TestMiddleware instead of beforeMiddleware with the before and after callbacks?